### DISCUSSION TITLE
Morphir cli tools as plugins.

### DISCUSSION THREAD

**Discussion by nwokafor-choongsaeng** [finos/morphir#140] [Ideas]
## Introducing CLI tools as Plugins.

The *morphir-elm* cli tool provides a host of useful functionalities including commands to build a Morphir project, generate code from an existing morphir project, run test cases, gather Morphir features used in a model, etc.
This discussion introduces an idea to split-up these *functionalities* as their own separate but pluggable/reusable tooling. 
Suggestions on ways to improve this solution, and even entirely different solutions are welcome.

## The Motivation

One of the evidence of growth of Morphir is in the size of it's ecosystem and available tooling, and as it leans more into the contribution of it's open-source community it becomes important that existing tools are, as much as possible, reusable in building new tooling.
The *morphir-elm* cli tool currently does not support a way to run a new backend on a morphir model and this means that, for every new backend, a new cli tool would have to be created that does most of what the *morphir-elm* tool already does i.e.

- reading a *morphir-ir.json*,
- decoding the Morphir distribution
- reading the *morphir.json*,
- reading additional configuration options,
- writing the generated sources to the file system.

This increases the development time, and can negatively impact the development experience of a contributor.

## Tooling as a Plugin

To solve the problem of reusability, the *morphir-elm* cli tool can be split up into *cli processing* and *morphir/IR processing* tools. The cli processing includes CLI specific tasks, like *initializing a morphir project*, *running a configured morphir tool*, reading and writing files and sources. This idea works well in conjunction with [splitting the morphir-elm npm package](https://github.com/finos/morphir/discussions/139) where the cli can be published under an npm package *@morphir/cli*, and the plugins (like the Scala backend) would be published as *@morphir/scala-backend-plugin*

## Improving the developer experience

With this change, developers can build their backends following a specific guideline for *building a Morphir plugin*, and specify that plugin as part of the Morphir configuration `morphir.json`. Allowing the morphir-cli to use do all the generic stuff (like reading and verifying inputs) and calling the specified backend with everything it needs to work based on what was configured.

**Comment by stephengoldbaum** [comment]
A plugin architecture for adding CLI features would be a big win.  A couple of thoughts:

- Are there existing frameworks that we can leverage?
- Right now, we're very npm-centric.  I wonder if it can be cross-language as well. I believe @DamianReeves has some thoughts on this.

**Reply by DamianReeves** [reply]
I'm hoping to dig a bit more into this and am actively thinking about this as I'm working on `morphir-dotnet`.

I'll try and add some comments here in the next few days with my thoughts.

**Reply by AttilaMihaly** [reply]
I'm really interested in finding something cross-language as well. The reason we are using NPM right now is that even though it comes from the JavaScript world it still seems to be the least language specific of all the package managers we looked at. It would be great if we could find something even less language specific.
